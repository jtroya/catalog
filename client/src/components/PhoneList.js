import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Row, Col, Card, CardTitle } from 'react-materialize';
import { Link } from 'react-router-dom';

import Loading from './Loading';

class PhoneList extends Component {
  componentDidMount() {
    this.props.fetchCatalog();
  }
  
  handleLinkDetail = (goTo, value) => {
    return (
      <Link to={ goTo }>{ value }</Link>
    );
  };

  displayHeader = (value) => {
    return (
      <CardTitle className="card_img" image={value}></CardTitle>
    );
  };

  renderList () {
    const { catalog } = this.props;
    if (catalog) {
      return catalog.map((item, index) => {
        return (
          <Col s={ 12 } m={ 6 } l={ 4 } key={ index }>
            <Card horizontal className="hoverable"
              header={ this.displayHeader(item.imageUrl)}
              title={ item.model}>
              { this.handleLinkDetail(`/detail/${item.phoneId}`, item.description) }
            </Card>
          </Col>
        );
      });
    }
  }
  
  render() {
    const { loading } = this.props;
    return (
      <div className="container">
        <Row>
          { loading &&
            <Loading size='big' /> }
          { !loading &&
            this.renderList()
          }
        </Row>
      </div>
    )
  }
}

PhoneList.propTypes = {
  loading: PropTypes.bool,
  catalog: PropTypes.array
};

export default PhoneList;
