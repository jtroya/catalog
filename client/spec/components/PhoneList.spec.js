import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MemoryRouter } from 'react-router-dom';

import PhoneList from '../../src/components/PhoneList';

describe('PhoneList', () => {
  let wrapper;
  const fetchCatalog = jest.fn();
  const fakeCatalog = [
    { model: 'my model', description: 'my description', imageUrl: 'img-urg', price: 123 }
  ];

  beforeEach(() => {
    wrapper = shallow(
      <MemoryRouter>
        <PhoneList 
          catalog={ fakeCatalog } 
          loading={ false } 
          fetchCatalog={ fetchCatalog } />
      </MemoryRouter>
    );
  });

  it('should render the component', () => {
    expect(wrapper.find('PhoneList')).toBeTruthy();
  });

  it('should render Loading component', () => {
    wrapper.setProps({ loading: true });
    expect(wrapper.find('Loading')).toBeTruthy();
    expect(wrapper.find('Col').length).toBe(0);
  });

  it('matches the snapshot', () => {
    const tree = shallow(<PhoneList 
      catalog={ fakeCatalog } 
      loading={ false } 
      fetchCatalog={ fetchCatalog } />);
    expect(toJson(tree)).toMatchSnapshot();
  });
});
