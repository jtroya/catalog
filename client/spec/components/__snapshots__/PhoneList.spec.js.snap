// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneList matches the snapshot 1`] = `
<div
  className="container"
>
  <Row>
    <Col
      key="0"
      l={4}
      m={6}
      s={12}
    >
      <Card
        className="hoverable"
        header={
          <CardTitle
            className="card_img"
            image="img-urg"
          />
        }
        horizontal={true}
        title="my model"
      >
        <Link
          replace={false}
          to="/detail/undefined"
        >
          my description
        </Link>
      </Card>
    </Col>
  </Row>
</div>
`;
